doc/BUILD.md

```diff
@@ -1,80 +1,109 @@
-Build Instructions
-==================
+# Build Instructions 
 
-## 1. Platform
+## Overview
+
+The instructions to build **Navigation2 Stack** and **Navigation2 Dependencies**.
+<br>
+
+  - [Build Options](#build-options)
+  - [Tutorials](#tutorials)
+  - [Troubleshooting  Guide](#troubleshooting-guide)
+
+## Recommended Platform
 * Ubuntu 18.04
 
-## 2. Manual Build Steps
-* The instructions are to build **Navigation2** package and its dependency packages.
-* To build **Navigation2** package from script, kindly refer [QuickStart](BUILD.md#quickstart-using-initial-setup-script).
-* To build **Navigation2** package for `master` branch, select the latest ROS2 distro in place of `ros2-distro-name` during `rosdep install`.
-* To build **Navigation2** package for specific distro branch (e.g- dashing-devel), select the matching ROS2 distro in place of `ros2-distro-name` during `rosdep install`.
+---
 
+## Build Options
 
-### 2.1 ROS2
-* Install ROS2 with dependencies from the ROS2 Installation Page: https://index.ros.org/doc/ros2/Installation/
-* Install the latest distribution of ROS2 to support Navigation2 `master`.
-* ROS2 can be installed either from binary packages or build from source. Choose to build ROS2 from source to support Navigation2 `master`.
+There two different recommended methods for building Navigation2 stack from source.
 
-### 2.2 Navigation2 Dependencies
-If ROS2 installed from source (Ignore if installed from binary packages)
-```console
-$ source ~/ros2_ws/install/setup.bash
-```
-Fetch, build and install navigation2 dependencies:
-```console
-$ mkdir -p ~/ros2_nav_dependencies_ws/src
-$ cd ~/ros2_nav_dependencies_ws
-$ wget https://raw.githubusercontent.com/ros-planning/navigation2/master/tools/ros2_dependencies.repos
-$ vcs import src < ros2_dependencies.repos
-$ rosdep install -y -r -q --from-paths src --ignore-src --rosdistro <ros2-distro-name>
-e.g- ros2-distro-name = dashing
-$ colcon build --symlink-install
-```
+### 1- Easy Build
+* An automated method to build Navigation2 in a simple and easy way.
+* To build Navigation2 by using the build script, go to [QuickStart](#2-quickstart-using-initial-setup-script)
```

Comment:
I'd like to get rid of the initial setup script. We don't test or maintain it really. It encodes the ROS2 build instructions which change too often.

Instead, I think there should be two types of build instructions.

1. Build from source on the release branch which has only a few steps:
   1. Install ROS2.
   2. Git clone our repo
   3. run rosdep
   4. colcon build
2. Build the master branch which has these steps.
   1. Install ROS2 master
   2. Make nav2 workspace
   3. Git clone nav2
   2. Make dependencies workspace
   3. vcs import ros2_dependencies
   4. rosdep dependencies
   5. colcon build dependencies
   6. source dependencies
   7 rosdep nav2
   8 colcon build nav2

I'd prefer to have these as two separate how-tos in the docs.

This readme is already too long in my opinion. As soon as I quickly scroll the document and see it goes on for screenfuls, I start thinking of other things I'd rather be doing.

Maybe to keep it simple, we could rewrite initial setup script to do steps 2 through 8 of the master build, and then point people to that script if they want to build master branch. The script could be part of the docs. It could be well commented and we could link to it from the docs.

Then we could have a how-to that describes how to just build navigation if you already have things setup. ie.
1. Updating navigation2
    1. Source nav2/setup.bash
    2. colcon build

and we can say, if any dependencies have changed you may have to go through some or all of the steps in the script to get them updated. We can leave that as an excercise for the reader.

Reply:
I use the init workspace script often

Reply:
Would you be ok with removing the ROS2 installation part of the script? That part seems the most difficult to maintain due to the frequently changing instructions.

Reply:
What ROS2 installation? I just see downloading of the nav2 stack, the deps workspace, and the ros2 workspace.

Reply:
https://github.com/ros-planning/navigation2/blob/5fe111a9d65d4d64d2f1f328559b21def159655e/tools/initial_ros_setup.sh#L55-L62

and 

https://github.com/ros-planning/navigation2/blob/5fe111a9d65d4d64d2f1f328559b21def159655e/tools/initial_ros_setup.sh#L86-L88

there more in the build_all.sh script.

The rosdep instructions in the script are wrong if they get run on the the ros2_ws. 
We have
```
rosdep install -y -r -q --from-paths . --ignore-src --rosdistro $ROS2_DISTRO --skip-keys "catkin"
```

The official build instructions, as of today say
```
rosdep install --from-paths src --ignore-src --rosdistro dashing -y --skip-keys "console_bridge fastcdr fastrtps libopensplice67 libopensplice69 rti-connext-dds-5.3.1 urdfdom_headers"
```

There's a large difference in the skip keys between the two.

Also there are a bunch of dependencies that are supposed be installed before doing the build that we don't cover in this file.

------------------------

If we keep the script, I'd like to get rid of the release branch support, so it only applies to the master branch.

For the master branch, well, we could keep it still if useful, so long as we add the caveat that the user is expected to read the source to understand what's been done, because it doesn't solve all problems, does more work than they want most of the time, and frequently gets out of date with the ROS 2 official build instructions.

Reply:
Those 2 lines aren't as different as you think. There's no change required. 

The point of the script is for a clean environment for all of master on ROS, the deps, and nav2. So I think absolutely it should be there.

I'm OK with just master branch. But I don't see where that really changes anything.